examples/conversation_with_codeact_agent/codeact_agent.py

```diff
@@ -0,0 +1,141 @@
+You should stop <execute> and provide an answer when they have already obtained the answer from the execution result. Whenever possible, execute the code for the user using <execute> instead of providing it.
+Your response should be concise, but do express their thoughts. Always write the code in <execute> block to execute them.
+You should not ask for the user's input unless necessary. Solve the task on your own and leave no unanswered questions behind.
+You should do every thing by your self.
+"""  # noqa
+
+EXAMPLE_MESSAGE = """
+Additionally, you are provided with the following code available:
+{example_code}
+The above code is already available in your interactive Python (Jupyter Notebook) environment, allowing you to directly use these variables and functions without needing to redeclare them.
+"""  # noqa
+
+
+class CodeActAgent(AgentBase):
+    """
+    The implementation of CodeAct-agent.
+    The agent can execute code interactively as actions.
+    More details can be found at the paper of codeact agent
+    https://arxiv.org/abs/2402.01030
+    and the original repo of codeact https://github.com/xingyaoww/code-act
+    """
+
+    def __init__(
+        self,
+        name: str,
+        model_config_name: str,
+        example_code: str = "",
+    ) -> None:
+        """ """
+        super().__init__(
+            name=name,
+            model_config_name=model_config_name,
+        )
+        self.n_max_executions = 5
+        self.example_code = example_code
+        self.code_executor = NoteBookExecutor()
+
+        sys_msg = Msg(name="system", role="system", content=SYSTEM_MESSAGE)
+        example_msg = Msg(
+            name="user",
+            role="user",
+            content=EXAMPLE_MESSAGE.format(example_code=self.example_code),
+        )
+
+        self.memory.add(sys_msg)
+
+        if self.example_code != "":
+            code_execution_result = self.code_executor.run_code_on_notebook(
+                self.example_code,
+            )
+            code_exec_msg = self.handle_code_result(
+                code_execution_result,
+                "Example Code excuted: ",
+            )
+            self.memory.add(example_msg)
+            self.memory.add(code_exec_msg)
+            self.speak(code_exec_msg)
+
+        self.parser = RegexTaggedContentParser(try_parse_json=False)
+
+    def handle_code_result(
+        self,
+        code_execution_result: ServiceResponse,
+        content_pre_sring: str = "",
+    ) -> Msg:
+        """return the message from code result"""
+        code_exec_content = content_pre_sring
+        if code_execution_result.status == ServiceExecStatus.SUCCESS:
+            code_exec_content += "Excution Successful:\n"
+        else:
+            code_exec_content += "Excution Failed:\n"
+        code_exec_content += "Execution Output:\n" + str(
+            code_execution_result.content,
+        )
+        return Msg(name="user", role="user", content=code_exec_content)
```

Comment:
"system" maybe better for the role field?

Reply:
As said above.